test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailNonPartitionKeyRestrictedQueriesTest.java

```diff
@@ -0,0 +1,202 @@
+        assertLogsOnAction(() -> execute("SELECT * from %s WHERE v1 = " + valueToQuery2), true, false, 1);
+    }
+
+    private void enableGuardrail()
+    {
+        cluster.forEach(instance -> instance.runOnInstance((IIsolatedExecutor.SerializableRunnable) () -> Guardrails.instance.setNonPartitionRestrictedQueryEnabled(true)));
+        assertGuardrailState(true);
+    }
+
+    private void disableGuardrail()
+    {
+        cluster.forEach(instance -> instance.runOnInstance((IIsolatedExecutor.SerializableRunnable) () -> Guardrails.instance.setNonPartitionRestrictedQueryEnabled(false)));
+        assertGuardrailState(false);
+    }
+
+    private void prepareSchema()
+    {
+        schemaChange("CREATE TABLE %s (k bigint, c bigint, v1 bigint, v2 bigint, PRIMARY KEY (k, c)) " +
+                     "WITH compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'};");
+
+        schemaChange("CREATE CUSTOM INDEX v1_sai_idx ON %s (v1)");
+        schemaChange("CREATE CUSTOM INDEX v2_sai_idx ON %s (v2)");
+
+        for (int i = 1; i < 4; i++)
+        {
+            cluster.get(i).acceptsOnInstance((IIsolatedExecutor.SerializableConsumer<String>) (t) -> {
+                for (ColumnFamilyStore cs : Keyspace.open("distributed_test_keyspace").getColumnFamilyStores())
+                {
+                    if (cs.name.equals(t))
+                    {
+                        cs.disableAutoCompaction();
+                    }
+                }
+            }).accept(tableName);
+        }
+
+        for (int i = 1; i < 4; i++)
+        {
+            Assert.assertTrue(cluster.get(i).applyOnInstance((IIsolatedExecutor.SerializableFunction<String, Boolean>) (t) ->
+            {
+                for (ColumnFamilyStore cs : Keyspace.open("distributed_test_keyspace").getColumnFamilyStores())
+                {
+                    if (cs.name.equals(t))
+                    {
+                        return cs.isAutoCompactionDisabled();
+                    }
+                }
+
+                return false;
+            }, tableName));
+        }
+    }
+
+    private long createSSTables(int numberOfSSTables, int... nodesToFlush)
+    {
+        Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
+
+        long k = System.currentTimeMillis();
+        long value = k + 10;
+        for (int i = 0; i < numberOfSSTables; i++)
+        {
+            // this will be replicated to each node
+            execute("INSERT INTO %s (k, c, v1, v2) VALUES (?, ?, ?, ?)", k, k, value, value + 10);
+
+            if (nodesToFlush.length == 0)
+                cluster.forEach((instance) -> instance.flush(KEYSPACE));
+            else
+                cluster.get(nodesToFlush).forEach((instance) -> instance.flush(KEYSPACE));
+        }
+
+        return value;
+    }
+
+    private void execute(String query, Object... args)
+    {
+        cluster.coordinator(1).execute(format(query), ConsistencyLevel.ALL, args);
```

Comment:
While I don't think it's harmful to look at the logging side effects when these trigger, that can be brittle, and it looks like `assertLogsOnAction()` is only checking generally for ERROR or WARN, not with any qualifiers that relate to the guardrail?

Either way, the basic assertion we need to make in the warning case is on the warnings returned to the client. In the lines above, you can access them this way:

```
SimpleQueryResult result = cluster.coordinator(1).executeWithResult(format(query), ConsistencyLevel.ALL, result.warnings(); // assert on the contents of this
```

In other words, I'm fine w/ keeping the logs checking and tying the assertions directly to specific messages for the features we're testing, but it might make more sense to focus on the actual client warnings and errors.

Reply:
Sure, I think that checking warnings directly from the response is better. We might get a log while executing some logic on WARN or ERROR level which is not necessarily related to the feature we are testing. It will be more stable as well etc.